File: mp3_convert.rb
# mp3_convert.rb
require 'find'
require 'optparse'
require 'shellwords'

INPUT_EXTS = ['.wav', '.flac', '.ogg', '.m4a', '.aac', '.wma', '.aiff', '.mp3', '.m4b']
OUTPUT_EXTS = ['mp3', 'wav', 'flac', 'ogg', 'm4a', 'aac']

def check_ffmpeg
  system('ffmpeg -version > /dev/null 2>&1') or begin
    $stderr.puts "Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH."
    exit 1
  end
end

def build_ffmpeg_cmd(input, output, bitrate, channels, sample_rate, normalize, format)
  cmd = ['ffmpeg', '-i', input, '-y']
  cmd += ['-b:a', "#{bitrate}k"]
  cmd += ['-ac', channels.to_s] if channels && channels > 0
  cmd += ['-ar', sample_rate.to_s] if sample_rate && sample_rate > 0
  cmd += ['-af', 'loudnorm=I=-16:LRA=11:TP=-1.5'] if normalize
  codec = case format
  when 'wav' then 'pcm_s16le'
  when 'flac' then 'flac'
  when 'ogg' then 'libvorbis'
  when 'm4a', 'aac' then 'aac'
  else 'libmp3lame'
  end
  cmd += ['-c:a', codec, output]
  cmd
end

def convert_file(input, output, bitrate, channels, sample_rate, normalize, format)
  cmd = build_ffmpeg_cmd(input, output, bitrate, channels, sample_rate, normalize, format)
  puts "  Выполняется: #{cmd.join(' ')}"
  system(*cmd)
end

def find_audio_files(root, recursive)
  files = []
  if File.file?(root) && INPUT_EXTS.include?(File.extname(root).downcase)
    return [root]
  end
  return files unless File.directory?(root)
  if recursive
    Find.find(root) do |path|
      files << path if File.file?(path) && INPUT_EXTS.include?(File.extname(path).downcase)
    end
  else
    Dir.glob(File.join(root, '*')).each do |path|
      files << path if File.file?(path) && INPUT_EXTS.include?(File.extname(path).downcase)
    end
  end
  files
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby mp3_convert.rb <вход> [опции]"
  opts.on("--bitrate N", Integer, "Битрейт (32-320)") { |v| options[:bitrate] = v }
  opts.on("--channels N", Integer, "Каналы (1=моно, 2=стерео)") { |v| options[:channels] = v }
  opts.on("--sample-rate N", Integer, "Частота (кГц)") { |v| options[:sample_rate] = v }
  opts.on("--normalize", "Нормализация") { options[:normalize] = true }
  opts.on("--format F", "Выходной формат") { |v| options[:format] = v }
  opts.on("--output FILE|DIR", "Выходной файл или папка") { |v| options[:output] = v }
  opts.on("--recursive", "Рекурсивный обход") { options[:recursive] = true }
  opts.on("--formats", "Показать форматы") { options[:formats] = true }
end.parse!

if options[:formats]
  puts "Поддерживаемые входные форматы: #{INPUT_EXTS.join(', ')}"
  puts "Поддерживаемые выходные форматы: #{OUTPUT_EXTS.join(', ')}"
  exit
end

check_ffmpeg

source = ARGV[0]
unless source
  puts "Укажите входной файл или папку."
  exit 1
end

bitrate = options[:bitrate] || 192
channels = options[:channels]
sample_rate = options[:sample_rate]
normalize = options[:normalize] || false
format = options[:format] || 'mp3'
output = options[:output]
recursive = options[:recursive] || false

unless OUTPUT_EXTS.include?(format)
  puts "Неизвестный формат: #{format}"
  exit 1
end

files = find_audio_files(source, recursive)
if files.empty?
  puts "Не найдено аудиофайлов в #{source}"
  exit 1
end

out_dir = output && File.directory?(output) ? output : './converted'
Dir.mkdir(out_dir) unless Dir.exist?(out_dir)
total = files.size
puts "Найдено #{total} аудиофайлов."
success = 0

files.each_with_index do |input_file, idx|
  rel = Pathname.new(input_file).relative_path_from(Pathname.new(source)).to_s rescue File.basename(input_file)
  out_path = if output && !File.directory?(output)
    output
  else
    File.join(out_dir, File.basename(rel, '.*') + ".#{format}")
  end
  FileUtils.mkdir_p(File.dirname(out_path))
  puts "[#{idx+1}/#{total}] Конвертация #{input_file} -> #{out_path}"
  if convert_file(input_file, out_path, bitrate, channels, sample_rate, normalize, format)
    success += 1
  else
    puts "  Ошибка при конвертации #{input_file}"
  end
end
puts "Готово! Успешно: #{success}, Всего: #{total}"

File: mp3_convert.cs
// mp3_convert.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

class Mp3Convert
{
    private static readonly HashSet<string> InputExts = new HashSet<string>
        { ".wav", ".flac", ".ogg", ".m4a", ".aac", ".wma", ".aiff", ".mp3", ".m4b" };
    private static readonly HashSet<string> OutputExts = new HashSet<string>
        { "mp3", "wav", "flac", "ogg", "m4a", "aac" };

    static void CheckFFmpeg()
    {
        try
        {
            var psi = new ProcessStartInfo("ffmpeg", "-version") { RedirectStandardOutput = true };
            using (var p = Process.Start(psi)) { p.WaitForExit(); }
        }
        catch
        {
            Console.Error.WriteLine("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.");
            Environment.Exit(1);
        }
    }

    static string[] BuildFFmpegCmd(string input, string output, int bitrate, int channels, int sampleRate, bool normalize, string format)
    {
        var cmd = new List<string>();
        cmd.Add("ffmpeg"); cmd.Add("-i"); cmd.Add(input); cmd.Add("-y");
        cmd.Add("-b:a"); cmd.Add($"{bitrate}k");
        if (channels > 0) { cmd.Add("-ac"); cmd.Add(channels.ToString()); }
        if (sampleRate > 0) { cmd.Add("-ar"); cmd.Add(sampleRate.ToString()); }
        if (normalize) { cmd.Add("-af"); cmd.Add("loudnorm=I=-16:LRA=11:TP=-1.5"); }
        string codec = format switch
        {
            "wav" => "pcm_s16le",
            "flac" => "flac",
            "ogg" => "libvorbis",
            "m4a" or "aac" => "aac",
            _ => "libmp3lame"
        };
        cmd.Add("-c:a"); cmd.Add(codec); cmd.Add(output);
        return cmd.ToArray();
    }

    static bool ConvertFile(string input, string output, int bitrate, int channels, int sampleRate, bool normalize, string format)
    {
        var args = BuildFFmpegCmd(input, output, bitrate, channels, sampleRate, normalize, format);
        Console.WriteLine($"  Выполняется: {string.Join(" ", args)}");
        try
        {
            var psi = new ProcessStartInfo("ffmpeg", string.Join(" ", args.Skip(1)))
            {
                UseShellExecute = false,
                RedirectStandardError = true
            };
            using (var p = Process.Start(psi))
            {
                p.WaitForExit();
                return p.ExitCode == 0;
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"  Ошибка: {e.Message}");
            return false;
        }
    }

    static List<string> FindAudioFiles(string root, bool recursive)
    {
        var files = new List<string>();
        if (File.Exists(root))
        {
            var ext = Path.GetExtension(root).ToLower();
            if (InputExts.Contains(ext)) files.Add(root);
            return files;
        }
        if (!Directory.Exists(root)) return files;
        var option = recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
        files.AddRange(Directory.GetFiles(root, "*.*", option)
            .Where(f => InputExts.Contains(Path.GetExtension(f).ToLower())));
        return files;
    }

    static async Task Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.Error.WriteLine("Использование: dotnet run <вход> [--bitrate N] [--channels 1|2] [--sample-rate N] [--normalize] [--format FMT] [--output FILE|DIR] [--recursive] [--formats]");
            return;
        }
        if (args[0] == "--formats")
        {
            Console.WriteLine("Поддерживаемые входные форматы: " + string.Join(", ", InputExts));
            Console.WriteLine("Поддерживаемые выходные форматы: " + string.Join(", ", OutputExts));
            return;
        }

        CheckFFmpeg();

        string source = args[0];
        int bitrate = 192;
        int channels = 0;
        int sampleRate = 0;
        bool normalize = false;
        string format = "mp3";
        string output = null;
        bool recursive = false;

        for (int i=1; i<args.Length; i++)
        {
            switch (args[i])
            {
                case "--bitrate": if (i+1 < args.Length) bitrate = int.Parse(args[++i]); break;
                case "--channels": if (i+1 < args.Length) channels = int.Parse(args[++i]); break;
                case "--sample-rate": if (i+1 < args.Length) sampleRate = int.Parse(args[++i]); break;
                case "--normalize": normalize = true; break;
                case "--format": if (i+1 < args.Length) format = args[++i]; break;
                case "--output": if (i+1 < args.Length) output = args[++i]; break;
                case "--recursive": recursive = true; break;
            }
        }

        if (!OutputExts.Contains(format))
        {
            Console.Error.WriteLine($"Неизвестный формат: {format}");
            Environment.Exit(1);
        }

        var files = FindAudioFiles(source, recursive);
        if (files.Count == 0)
        {
            Console.WriteLine($"Не найдено аудиофайлов в {source}");
            return;
        }

        string outDir = output != null && Directory.Exists(output) ? output : "./converted";
        Directory.CreateDirectory(outDir);
        int total = files.Count;
        Console.WriteLine($"Найдено {total} аудиофайлов.");
        var semaphore = new SemaphoreSlim(4);
        var tasks = new List<Task>();
        int success = 0;

        for (int i=0; i<total; i++)
        {
            await semaphore.WaitAsync();
            int idx = i;
            string inputFile = files[i];
            tasks.Add(Task.Run(() =>
            {
                try
                {
                    var rel = Path.GetRelativePath(source, inputFile);
                    if (rel == ".") rel = Path.GetFileName(inputFile);
                    string outPath = Path.Combine(outDir, Path.ChangeExtension(rel, "." + format));
                    if (output != null && !Directory.Exists(output)) outPath = output;
                    Directory.CreateDirectory(Path.GetDirectoryName(outPath));
                    Console.WriteLine($"[{idx+1}/{total}] Конвертация {inputFile} -> {outPath}");
                    if (ConvertFile(inputFile, outPath, bitrate, channels, sampleRate, normalize, format))
                        Interlocked.Increment(ref success);
                }
                catch (Exception e)
                {
                    Console.WriteLine($"  Ошибка при конвертации {inputFile}: {e.Message}");
                }
                finally { semaphore.Release(); }
            }));
        }
        await Task.WhenAll(tasks);
        Console.WriteLine($"Готово! Успешно: {success}, Всего: {total}");
    }
}
